lib/src/client/transport/xhr_transport.dart

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2017, the gRPC project authors. Please see the AUTHORS file
+// for details. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:async';
+import 'dart:html';
+import 'dart:typed_data';
+
+import 'package:meta/meta.dart';
+
+import '../../shared/streams.dart';
+
+import '../options.dart';
+
+import 'transport.dart';
+
+class XhrTransportStream extends GrpcTransportStream {
+  HttpRequest _request;
+  int _requestBytesRead = 0;
+  StreamController<ByteBuffer> _incomingProcessor;
+  StreamController<GrpcMessage> _incomingMessages;
+  StreamController<List<int>> _outgoingMessages;
+
+  @override
+  Stream<GrpcMessage> get incomingMessages => _incomingMessages.stream;
+
+  @override
+  StreamSink<List<int>> get outgoingMessages => _outgoingMessages.sink;
+
+  XhrTransportStream(this._request) {
+    _incomingProcessor = new StreamController();
```

Comment:
Not sure why this has to be a stream-controller. Isn't it better as a transform?

Reply:
I'm going to have to go with your judgement there. If you think so, I agree and I'll switch it over.

Reply:
I looked into this (forgot it in the initial update) but because I need to add to the stram from onProgress, I think this still makes sense as a stream controller, even though it's piping out to a secondary stream.

It is possible I'm just not sure how to properly implement it as a transformer though.